pkg/storage/replica.go

```diff
@@ -4157,7 +4165,7 @@ func (r *Replica) quiesceAndNotifyLocked(ctx context.Context, status *raft.Statu
 			To:     id,
 			Type:   raftpb.MsgHeartbeat,
 			Term:   status.Term,
-			Commit: status.Commit,
+			Commit: min(prog.Match, status.Commit),
```

Comment:
Please add a comment such as this one:

// In common operation, we only quiesce when all followers are up-to-date, and
// so the status.Commit field is safe to send. However, when we quiesce in the
// presence of dead nodes, a follower which is behind and counted as dead may
// not have the log entry referenced by `status.Commit` and would explode if it
// were told to commit up to that point. So reduce the commit index to an index
// we know the follower has acknowledged.

Reply:
I've incorporated this into the new mechanism suggested by @bdarnell.